content/en/blog/_posts/2025-05-12-ensure-secret-pulled-images/index.md

```diff
@@ -0,0 +1,136 @@
+>}}
+
+While `IfNotPresent` should not pull *image Foo* if it is already present
+on the node, it is an incorrect security posture to allow all pods scheduled
+to a node to have access to previously pulled private image. These pods were never
+authorized to pull the image in the first place.
+
+## IfNotPresent, but only if I am supposed to have it
+
+In Kubernetes 1.33, we - SIG Auth and SIG Node - are finally addressing this (really old) problem and getting the verification right! The basic expected behavior is not changed. If
+an image is not present, the Kubelet will attempt to pull the image. The credentials each pod supplies will be utilized for this task. This matches behavior prior to 1.33.
+
+If the image is present, then the behavior of the Kubelet changes. The Kubelet will now
+verify the pod's credentials before allowing the pod to use the image.
+
+Performance and service stability have been a consideration while revising the feature.
+Pods utilizing the same credential will not be required to re-authenticate. This is
+also true when pods source credentials from the same Kubernetes Secret object, even
+when the credentials are rotated.
+
+## Never pull, but use if authorized
+
+The `imagePullPolicy: Never` option does not fetch images. However, if the
+container image is already present on the node, any pod attempting to use the private
+image will be required to provide credentials, and those credentials require verification.
+
+Pods utilizing the same credential will not be required to re-authenticate.
+Pods that do not supply credentials previously used to successfully pull an
+image will not be allowed to use the private image.
+
+## Always pull, if authorized
+
+The `imagePullPolicy: Always` has always worked as intended. Each time an image
+is requested, the request goes to the registry and the registry will perform an authentication
+check.
+
+In the past, forcing the `Always` image pull policy via pod admission was the only way to ensure
+that your private container images didn't get reused by other pods on nodes which already pulled the images.
+
+Fortunately, this was somewhat performant. Only the image manifest was pulled, not the image. However, there was still a cost and a risk. During a new rollout, scale up, or pod restart, the image registry that provided the image MUST be available for the auth check, putting the image registry in the critical path for stability of services running inside of the cluster.
+
+## How it all works
+
+The feature is based on persistent, file-based caches that are present on each of
+the nodes. The following is a simplified description of how the feature works.
+For the complete version, please see [KEP-2535](https://github.com/kubernetes/enhancements/blob/0262549808ebe7e64b94ee6f8e4fbdd024332470/keps/sig-node/2535-ensure-secret-pulled-images/README.md).
```

Comment:
```suggestion
For the complete version, please see [KEP-2535](https://kep.k8s.io/2535).
```

did you intend to pin this to a specific commit?

Reply:
Nope!  Appreciate that.